### PR TITLE
Support vending products that are backed by binaryTargets

### PR DESCRIPTION
**Pull request by neonichu**
This adds support for vending an executable product that consists solely of a binary target that is backed by an artifact bundle. This allows vending binary executables as their own separate package, independently of the plugins that are using them.

rdar://101096803

### CODE REVIEW

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
Some tests need updating due to new diagnostic, but this one surprises me a bit:
```
<EXPR>:0: error: PluginTests.testLocalAndRemoteToolDependencies : threw error "signalled(4): /home/build-user/swiftpm/.build/x86_64-unknown-linux-gnu/debug/swift-package --package-path /tmp/Miscellaneous_Plugins_PluginUsingLocalAndRemoteTool.QzzITS/MyLibrary --configuration debug plugin my-plugin output:
    Commands/SwiftPackageTool.swift:1083: Fatal error: no product named RemoteTool
```

File: Sources/PackageLoading/ManifestLoader+Validation.swift [hunk 1]
```diff
@@ -78,10 +78,14 @@ public struct ManifestValidator {
...
-            if areTargetsBinary && product.type != .library(.automatic) {
-                diagnostics.append(.invalidBinaryProductType(productName: product.name))
+            // Check that products that reference only binary targets don't define an explicit library type.
+            if product.targets.allSatisfy({ self.manifest.targetMap[$0]?.type == .binary }) {
+                switch product.type {
+                case .library(.automatic), .executable:
+                    break
+                default:
+                    diagnostics.append(.invalidBinaryProductType(productName: product.name))
+                }
```
Comment (FranzBusch) [inline Sources/PackageLoading/ManifestLoader+Validation.swift:88]: If I understand this correctly, this still allows products that have both a binary target and a source target to work right? Do we want to to allow this at all? What happens if a I have a library product that depends on a source target and a binaryTarget that is an executable?

Comment (neonichu) [reply Sources/PackageLoading/ManifestLoader+Validation.swift:88]: This is the validation for the executable product itself which can only vend a singular target which now can either be an executable source target or a binary target that contains an executable.

File: Sources/Build/BuildPlan.swift [hunk 2]
```diff
@@ -1989,8 +1989,8 @@ public class BuildPlan {
 
         var productMap: [ResolvedProduct: ProductBuildDescription] = [:]
         // Create product description for each product we have in the package graph except
-        // for automatic libraries and plugins, because they don't produce any output.
-        for product in graph.allProducts where product.type != .library(.automatic) && product.type != .plugin {
+        // for automatic libraries, plugins or products which consist solely of binary targets, because they don't produce any output.
+        for product in graph.allProducts where product.type != .library(.automatic) && product.type != .plugin && !product.targets.filter({ !($0.underlyingTarget is BinaryTarget) }).isEmpty {
```
Comment (neonichu) [inline Sources/Build/BuildPlan.swift:1993]: Somewhat interesting: binary-only libraries just happened to work because they're also automatic products :)

File: Sources/PackageModel/Target.swift [hunk 3]
```diff
@@ -638,6 +638,11 @@ public final class BinaryTarget: Target {
         }
     }
 
+    public var containsExecutable: Bool {
+        // FIXME: needs to be revisited once libraries are supported in artifact bundles
```
Comment (neonichu) [inline Sources/PackageModel/Target.swift:642]: I opted to not open this can of worms here since right now `.artifactsArchive` implies that there is an executable. Addressing this required too many changes that are completely unrelated to this PR and which should be done once we implement support for libraries in artifact bundles.

File: Sources/SPMBuildCore/PluginInvocation.swift [hunk 4]
```diff
@@ -502,29 +502,31 @@ public extension PluginTarget {
...
-                    // TODO: Memoize this result for the host triple
-                    guard let execInfos = try? target.parseArtifactArchives(for: hostTriple, fileSystem: fileSystem) else {
-                        // TODO: Deal better with errors in parsing the artifacts
-                        return []
-                    }
-                    return execInfos.map{ .vendedTool(name: $0.name, path: $0.executablePath) }
-                }
-                // For an executable target we create a `builtTool`.
-                else if target.type == .executable {
-                    // TODO: How do we determine what the executable name will be for the host platform?
```
Comment (neonichu) [inline Sources/SPMBuildCore/PluginInvocation.swift:520]: Note: I deleted this TODO because supporting the case where plugins work when the target platform isn't the host requires significant work in SwiftPM's build system.

File: Tests/PackageLoadingTests/PD_5_3_LoadingTests.swift [hunk 5]
```diff
@@ -177,29 +177,7 @@ class PackageDescription5_3LoadingTests: PackageDescriptionLoadingTests {
...
-                        .binaryTarget(name: "Foo", path: "Foo.xcframework"),
-                    ]
-                )
-                """
-
-            let observability = ObservabilitySystem.makeForTesting()
-            let (_, validationDiagnostics) = try loadAndValidateManifest(content, observabilityScope: observability.topScope)
-            XCTAssertNoDiagnostics(observability.diagnostics)
-            testDiagnostics(validationDiagnostics) { result in
-                result.check(diagnostic: "invalid type for binary product 'FooLibrary'; products referencing only binary targets must have a type of 'library'", severity: .error)
```
Comment (neonichu) [inline Tests/PackageLoadingTests/PD_5_3_LoadingTests.swift:202]: This diagnostic can longer be emitted for this particular case, because we cannot know whether a binary target is vending an executable or a library during manifest parsing since we need to crack open the referenced artifact to determine that. There will still be a diagnostic for this case, but it is handled by `PackageBuilder` and already covered by other existing tests.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

File: Fixtures/Miscellaneous/Plugins/BinaryToolProductPlugin/Dependency/.gitignore [hunk 6]
```diff
@@ -0,0 +1,9 @@
+.DS_Store
+/.build
+/Packages
+/*.xcodeproj
+xcuserdata/
+DerivedData/
+.swiftpm/config/registries.json
+.swiftpm/xcode/package.xcworkspace/contents.xcworkspacedata
+.netrc
```
Comment (tomerd) [inline Fixtures/Miscellaneous/Plugins/BinaryToolProductPlugin/Dependency/.gitignore:9]: redundant?

File: Sources/Build/BuildPlan.swift (hunk 2 above)
Comment (tomerd) [reply Sources/Build/BuildPlan.swift:1993]: nit: since the condition here is starting to become complex, maybe a private extension on product that makes it more readable

File: Sources/PackageModel/Target.swift (hunk 3 above)
Comment (tomerd) [reply Sources/PackageModel/Target.swift:642]: I think this is fair, but would be nice if there was some kind of associated data on the artifactsArchive that would allow us to throw an error if it is not an executable. eg `.artifactsArchive(type:)` not sure if that is the can of warms tho :)

Comment (neonichu) [reply Sources/PackageModel/Target.swift:642]: Yah, that was my approach, but it did require a bunch of changes 🪱  Could be a good change for a separate PR, though.

**Review by tomerd — COMMENTED** [review]
overall looks good. some small suggestions. would be good to get @abertelrud review as he knows this part of the code best

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Review by abertelrud — APPROVED** [review]
